src/Compression/CompressionCodecEncrypted.cpp

```diff
@@ -11,204 +19,291 @@
-    }
+String CompressionCodecEncrypted::Configuration::getCurrentKey() const
+{
+    if (!params.get())
+        throw Exception("Empty params in CompressionCodecEncrypted configuration", ErrorCodes::BAD_ARGUMENTS);
+    return params.get()->keys_storage.at(params.get()->current_key_id);
+}
 
-    std::string CompressionCodecEncrypted::deriveKey(const std::string_view & master_key)
-    {
-        std::string_view salt(""); // No salt: derive keys in a deterministic manner.
-        std::string_view info("Codec Encrypted('AES-128-GCM-SIV') key generation key");
-        std::array<char, 32> result;
-
-        const int ok = HKDF(reinterpret_cast<uint8_t *>(result.data()), result.size(),
-                            EVP_sha256(),
-                            reinterpret_cast<const uint8_t *>(master_key.data()), master_key.size(),
-                            reinterpret_cast<const uint8_t *>(salt.data()), salt.size(),
-                            reinterpret_cast<const uint8_t *>(info.data()), info.size());
-        if (!ok)
-            throw Exception(lastErrorString(), ErrorCodes::OPENSSL_ERROR);
-
-        return std::string(result.data(), 16);
-    }
+UInt64 CompressionCodecEncrypted::Configuration::getCurrentID() const
+{
+    if (!params.get())
+        throw Exception("Empty params in CompressionCodecEncrypted configuration", ErrorCodes::BAD_ARGUMENTS);
+    return params.get()->current_key_id;
+}
+
+String CompressionCodecEncrypted::Configuration::getNonce() const
+{
+    if (!params.get())
+        throw Exception("Empty params in CompressionCodecEncrypted configuration", ErrorCodes::BAD_ARGUMENTS);
+    return params.get()->nonce;
+}
+
+CompressionCodecEncrypted::CompressionCodecEncrypted()
+{
+    setCodecDescription("AES_128_GCM_SIV");
+}
+
+uint8_t CompressionCodecEncrypted::getMethodByte() const
+{
+    return static_cast<uint8_t>(CompressionMethodByte::Encrypted);
+}
+
+void CompressionCodecEncrypted::updateHash(SipHash & hash) const
+{
+    getCodecDesc()->updateTreeHash(hash);
+}
+
+UInt32 CompressionCodecEncrypted::getMaxCompressedDataSize(UInt32 uncompressed_size) const
+{
+    // The GCM mode is a stream cipher. No paddings are
+    // involved. There will be a tag at the end of ciphertext (16
+    // octets).
+    return uncompressed_size + 24;
+}
+
+UInt32 CompressionCodecEncrypted::doCompressData(const char * source, UInt32 source_size, char * dest) const
+{
+    // Generate an IV out of the data block and the key-generation
+    // key. It is completely deterministic, but does not leak any
+    // information about the data block except for equivalence of
+    // identical blocks (under the same master key). The IV will
+    // be used as an authentication tag. The ciphertext and the
+    // tag will be written directly in the dest buffer.
+    const std::string_view plaintext = std::string_view(source, source_size);
+
+    encrypt(plaintext, dest);
+    return source_size + 24;
```

Comment:
Seems to be wrong. I suppose `doCompressData()` should return something like `source_size + tag_size + varint_size`, where `varint_size` can be any value between 1 and 8.